lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+// orders the transactions by smallest timestamp. In case of a tie, transactions are ordered by greatest fee rate. Finally,
+// in case of another tie, transactions are ordered by their hash.
+func TimeBucketComparator(a, b interface{}) int {
+	aVal := a.(*MempoolTx)
+	bVal := b.(*MempoolTx)
+
+	if aVal.Added.UnixNano() > bVal.Added.UnixNano() {
+		return 1
+	} else if aVal.Added.UnixNano() < bVal.Added.UnixNano() {
+		return -1
+	} else {
+		if aVal.FeePerKB < bVal.FeePerKB {
+			return 1
+		} else if aVal.FeePerKB > bVal.FeePerKB {
+			return -1
+		} else {
+			return HashToBigint(aVal.Hash).Cmp(HashToBigint(bVal.Hash))
+		}
+	}
+}
+
+// AddTransaction adds a transaction to the TimeBucket. It returns an error if the transaction is outside the
+// TimeBucket's fee range, or if the transaction hash is nil.
+func (tb *TimeBucket) AddTransaction(txn *MempoolTx) error {
+	if txn.Hash == nil {
+		return fmt.Errorf("transaction hash is nil")
+	}
+
+	if tb.feeMin > txn.FeePerKB || tb.feeMax < txn.FeePerKB {
+		return fmt.Errorf("transaction fee %d outside of bucket range [%d, %d]", txn.FeePerKB, tb.feeMin, tb.feeMax)
+	}
+
+	tb.txnsSet.Add(txn)
+	return nil
+}
+
+// RemoveTransaction removes a transaction from the TimeBucket.
+func (tb *TimeBucket) RemoveTransaction(txn *MempoolTx) {
+	tb.txnsSet.Remove(txn)
+}
+
+func (tb *TimeBucket) Empty() bool {
+	return tb.txnsSet.Empty()
+}
+
+// GetIterator returns an iterator over the MempoolTx inside the TimeBucket.
+func (tb *TimeBucket) GetIterator() treeset.Iterator {
+	return tb.txnsSet.Iterator()
+}
+
+// GetTransactions returns a slice of MempoolTx inside the TimeBucket. The slice is ordered according to the
+// TimeBucketComparator.
+func (tb *TimeBucket) GetTransactions() []*MempoolTx {
+	txns := []*MempoolTx{}
+	it := tb.GetIterator()
+	for it.Next() {
+		if txn, ok := it.Value().(*MempoolTx); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+//============================================
+//	Fee-Time Bucket Math
+//============================================
+
+// computeFeeBucketBaseAndMultiplierFromGlobalParams takes the fee base rate and multiplier for the GlobalParamsEntry,
+// and returns them as big.Floats.
+func computeFeeBucketBaseAndMultiplierFromGlobalParams(globalParams *GlobalParamsEntry) (
+	_baseRate *big.Float, _bucketMultiplier *big.Float) {
+
+	feeBucketBaseRate := NewFloat().SetUint64(globalParams.FeeBucketBaseRate)
+	feeBucketMultiplier := NewFloat().SetUint64(10000 + globalParams.FeeBucketMultiplierBasisPoints)
+	feeBucketMultiplier.Quo(feeBucketMultiplier, NewFloat().SetUint64(10000))
+	return feeBucketBaseRate, feeBucketMultiplier
+}
+
+// computeTimeBucketMinFromIndex takes a fee range index and the GlobalParamsEntry, and returns the minimum fee rate
+// of this fee range.
+func computeTimeBucketMinFromIndex(index int, globalParams *GlobalParamsEntry) uint64 {
```

Comment:
No need to pass in the entire global params here. The caller can extract just the required params and pass them in directly. This way, the function signature becomes self documenting re: what data it actually needs. 
```suggestion
func computeTimeBucketMinFromIndex(index uint64, feeBucketBaseRateNanos *big.Float, feeBucketMultiplier *big.Float) uint64 {
```